### PR TITLE
Omit loop restoration on fully skipped areas

### PR DESCRIPTION
**Pull request by KyleSiefring**

### CODE REVIEW

**Comment by KyleSiefring** [comment]
Need to add stats and speedups to commit message.

File: src/rdo.rs [hunk 1]
```diff
@@ -2066,6 +2066,90 @@ pub fn rdo_loop_decision<T: Pixel, W: Writer>(
...
+          }
+        }
+        cdef_skip[sby * MAX_SB_SIZE + sbx] = skip;
+        cdef_skip_all &= skip;
+      }
+    }
+  }
+
+  // Unlike cdef, loop restoration will run regardless of whether blocks are
+  // skipped or not. At the same time, the most significant improvement wil
```
Comment (lu-zero) [inline src/rdo.rs:2102]:
```suggestion
  // skipped or not. At the same time, the most significant improvement will
```

**Comment by KyleSiefring** [comment]
| Speed       | 2      | 3      | 4      | 5      | 6      | 9      | 10     |
|-------------|--------|--------|--------|--------|--------|--------|--------|
| Encode Time | -8.51  | -16.41 | -20.02 | -17.38 | -15.89 | -18.84 | 1.55   |
| PSNR Y      | 0.1520 | 0.1869 | 0.1438 | 0.1615 | 0.1951 | 0.5331 | 0.0000 |
| PSNR Cb     | 0.4569 | 0.5378 | 0.5551 | 0.5473 | 0.4967 | 0.7636 | 0.0000 |
| PSNR Cr     | 0.7728 | 1.0823 | 1.0894 | 1.0539 | 0.7348 | 1.2074 | 0.0000 |
| CIEDE2000   | 0.3322 | 0.3863 | 0.4031 | 0.3646 | 0.3963 | 0.8263 | 0.0000 |
| SSIM        | 0.1735 | 0.2440 | 0.2132 | 0.2012 | 0.2066 | 0.6623 | 0.0000 |
| MS-SSIM     | 0.1646 | 0.2419 | 0.2114 | 0.2082 | 0.2496 | 0.7791 | 0.0000 |
| PSNR-HVS Y  | 0.1307 | 0.1724 | 0.1226 | 0.1504 | 0.1264 | 0.4689 | 0.0000 |
| PSNR-HVS Cb | 0.2266 | 0.3721 | 0.4658 | 0.4165 | 0.3825 | 0.5781 | 0.0000 |
| PSNR-HVS Cr | 0.3649 | 0.7772 | 0.6678 | 0.6689 | 0.4635 | 0.7829 | 0.0000 |
| PSNR-HVS    | 0.1385 | 0.1878 | 0.1393 | 0.1669 | 0.1374 | 0.4690 | 0.0000 |
| VMAF        | 0.1090 | 0.3370 | 0.2963 | 0.2060 | 0.1980 | 0.5031 | 0.0000 |
| VMAF-NEG    | 0.1434 | 0.2712 | 0.2465 | 0.2171 | 0.1321 | 0.3654 | 0.0000 |

**Comment by shssoichiro** [comment]
I think the speed-to-cost tradeoff is pretty worthwhile. Is it worth making a speed feature out of this? Maybe we want to disable it at slower speed levels (0, 1, maybe 2?)

**Review by shssoichiro — APPROVED** [review]

**Comment by KyleSiefring** [comment]
Currently seeing unexpected changes on speed 0 vs master.

**Comment by KyleSiefring** [comment]
Fixed it. New, improved stats.

| Speed       | 2       | 3      | 4       | 5      | 6       | 9       | 10     |
|-------------|---------|--------|---------|--------|---------|---------|--------|
| Encode Time | -9.28   | -15.48 | -21.35  | -9.61  | -14.96  | -16.09  | 3.23   |
| PSNR Y      | 0.0821  | 0.1242 | 0.0745  | 0.0715 | 0.1260  | 0.4148  | 0.0000 |
| PSNR Cb     | -0.0030 | 0.0440 | -0.0926 | 0.0646 | 0.0457  | -0.0562 | 0.0000 |
| PSNR Cr     | 0.0079  | 0.0528 | 0.0347  | 0.1480 | 0.0708  | 0.0906  | 0.0000 |
| CIEDE2000   | 0.0969  | 0.1566 | 0.1000  | 0.1556 | 0.1798  | 0.4378  | 0.0000 |
| SSIM        | 0.0555  | 0.1075 | 0.0506  | 0.0481 | 0.0617  | 0.3609  | 0.0000 |
| MS-SSIM     | 0.0978  | 0.1514 | 0.1166  | 0.1176 | 0.1823  | 0.5675  | 0.0000 |
| PSNR-HVS Y  | 0.0765  | 0.1135 | 0.0503  | 0.0738 | 0.1012  | 0.3501  | 0.0000 |
| PSNR-HVS Cb | -0.0669 | 0.0353 | -0.0970 | 0.0892 | 0.0597  | -0.0295 | 0.0000 |
| PSNR-HVS Cr | -0.0815 | 0.0102 | -0.0177 | 0.1484 | -0.0022 | 0.0123  | 0.0000 |
| PSNR-HVS    | 0.0770  | 0.1183 | 0.0539  | 0.0833 | 0.1039  | 0.3368  | 0.0000 |
| VMAF        | 0.1658  | 0.2535 | 0.2528  | 0.2053 | 0.2131  | 0.4359  | 0.0000 |
| VMAF-NEG    | 0.1559  | 0.2322 | 0.2075  | 0.1916 | 0.1532  | 0.3008  | 0.0000 |

**Pull request merged**